datafusion/sqllogictest/test_files/order.slt

```diff
@@ -410,3 +410,38 @@ SELECT DISTINCT time as "first_seen" FROM t ORDER BY 1;
 ## Cleanup
 statement ok
 drop table t;
+
+# Create a table having 3 columns which are ordering equivalent by the source. In the next step,
+# we will expect to observe the removed sort exec by propagating the orders across projection.
+statement ok
+CREATE EXTERNAL TABLE multiple_ordered_table (
+  a0 INTEGER,
+  a INTEGER,
+  b INTEGER,
+  c INTEGER,
+  d INTEGER
+)
+STORED AS CSV
+WITH HEADER ROW
+WITH ORDER (a ASC)
+WITH ORDER (b ASC)
+WITH ORDER (c ASC)
+LOCATION '../core/tests/data/window_2.csv';
+
+query TT
+EXPLAIN SELECT (b+a+c) AS result 
+FROM multiple_ordered_table
+ORDER BY result;
+----
+logical_plan
+Sort: result ASC NULLS LAST
+--Projection: multiple_ordered_table.b + multiple_ordered_table.a + multiple_ordered_table.c AS result
+----TableScan: multiple_ordered_table projection=[a, b, c]
+physical_plan
+SortPreservingMergeExec: [result@0 ASC NULLS LAST]
```

Comment:
The `SortPreservingMergeExec` seems still rather "nonoptimal" as the batches themselves are already sorted (but the info is thrown away by `RepartitionExec`). Any plans for addressing this as well in the future?

Reply:
Yes, if I'm not mistaken one of the WIP PRs in our pipeline will address this 🙂

Reply:
RepartitionExec does not break the order of partitions in this case as it partitions 1 to 4. Because these partitions are already ordered, the presence of SortPreservingMerge is correct. For general cases, SortPreservingRepartitionExec implementation is on the way, which will have the capability to preserve order for all kinds of partitioning.

Reply:
What I essentially mean is this:

The batches of the table `multiple_ordered_table ` are ordered and order is preserved in `RepartitionExec` and `ProjectionExec`. If `SortPreservingMergeExec` would know the number of the batch (batch 0, batch 1, batch 2, etc.) it would only need to wait on batch 0, batch 1, batch 2, etc. to appear from the partition streams, but not the rows itself, which would be much faster.

Reply:
Is that something you plan as well?

Reply:
To maintain batch indices for preserving order, it’s actually a good idea. The current sort preserving algorithm, designed to preserve the hash repartition, tends to overfit to the row sorting. We could potentially collaborate on implementing this optimization in future work.